### PR TITLE
Table: Adds header rendering slot

### LINKED ISSUE #4908: [Question] Custom header for the el-table-column
Hello!

Could i have a custom template for the column header? 

I need to have some filtering using el-input.

Endresult should be similar to

screenshot_2017-05-17_13-13-53

### LINKED ISSUE #4909: [Feature Request] Custom header for the el-table-column
### Existing Component
Yes

### Component Name
el-table-column

### Description
Hello!

Could i have a custom template for the column header?

I need to have some filtering using el-input.

Endresult should be similar to

screenshot_2017-05-17_13-13-53

### PR DESCRIPTION
**Pull request by ivanseidel**
Fixes #4909 
Fixes #4908
Related #11807
Related #6573 

Now it is possible to assign custom headers directly to templates, exactly as renderHeader does.

Please make sure these boxes are checked before submitting your PR, thank you!

* [x] Make sure you follow Element's contributing guide ([中文](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.zh-CN.md) | [English](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.en-US.md) | [Español](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.es.md)).
* [x] Make sure you are merging your commits to `dev` branch.
* [x] Add some descriptions and refer relative issues for you PR.

### CODE REVIEW

**Comment by ziyoung** [comment]
TableColumns has `render-header` attribute which allows you custom column content. Is it necessary to add header named slot? If I set `render-header` and use header slot, how to render this column?

**Comment by ivanseidel** [comment]
Slots are more easy to create than render function, as you dont need to write any sort of component code, just template.

See how hard it is to write custom render functions for pretty simple things in #4909.

My initial idea was to create a scope and to create another check in table-header to render it, but then I noticed that the syntax of render-header could be used in the same way.

I would say that scoped slots should take precedence, just like a render-header takes precedence on simple label rendering. It isnt that different, as current behavior opts for only one of the methods.

As you can see in my code, this change doesnt adds much overhead, since the template is only being aplied to the renderHeader prop during mounting. The rest is just like it was supposed to be: a custom renderHeader function.

Also, if we agree that "renderHeader" should take precedence over slots (witch I think is not good in my opinion), is just a matter of checking for the prop and avoiding to set it. It can be possible in both ways...

**Comment by ziyoung** [comment]
@ivanseidel you can use vue-jsx to set `render-function`. It's not hard to create than render function.
https://codesandbox.io/s/w2jyqwjzql
 If there are two ways to custom column content, this will cause confusion.

**Comment by ivanseidel** [comment]
I dont think it will cause confusion. This should be discussed.

I spent over two hours to work with a render function in my controller, and I think I spent my time for nothing as I could easily have done that just like the docs says on the body funcion of the cell.

It doesnt make sense to let users decide their cell content on scoped slots but not the header. If that made sense it should be both render function or scoped slots, not one different from the other.

From the perspective of simplicity, it is much more easy to work with slots, also, it is The documented way on Vue docunentation to do so. Assigning functions as parameters to be called a single time is not a common scenario on vue, and should be a anti-pattern.

I suggest to get more people envolved, because I love Element and I see this is an easy inprovement with no conflicts that will make people life much easier to create custom tables.

Thankz

**Comment by ivanseidel** [comment]
Also, the main idea of separating Vue.js controller, style and template is to avoid having to work on layout, styling and animations in Controllers. 

Your example lacks on showing this working, since there is a method responsible for rendering elements. It should be avoided, to make code cleaner and more maintainable. This looks more like `React` than `Vuejs`. Maybe it is a good approach for `React`.

``
`
methods: {
    renderHeader(h, { column }) {
      return <el-button type="primary">{ column.label }</el-button>
    }
  }
```

**Comment by ziyoung** [comment]
@ivanseidel Yes. Compared to render function, slots are easier to use.
If we add the header named slot, there are two ways to custom table column header.
I'm reminded of this sentence in [The Zen of Python](https://www.python.org/dev/peps/pep-0020/).
> There should be one-- and preferably only one --obvious way to do it.

So, if header slot is supported, should we deprecate `render-header` function in next major version and recommend users to use header slot?

**Comment by ivanseidel** [comment]
For me we have 3 options to choose:

1. Add slots, and make it "priority" over renderHeader.

2. Add slots and make it "less priority" (will opt for renderHeader instead of slot)

3. Go with option 1, and log a warning of deprecation if user inputs anything in parameter (justlike you said).

I would go for option 3, as we would be able to "measure" how many people still require thia feature by showing the link of thiz pull request. If people really need renderHeader after all, we could just make it something ocluded from the docs, or less direct for users. They would then use primarilly slots and if they need, renderHeader.

What so you think?

**Comment by ziyoung** [comment]
@ivanseidel I agree with you.

File: examples/docs/en-US/table.md [hunk 1]
```diff
@@ -1500,6 +1501,79 @@ Customize table column so it can be integrated with other components.
 ```
 :::
 
+### Table with custom header
```
Comment (ziyoung) [inline examples/docs/en-US/table.md]: Update the missing Spanish and Chinese docs (You don't need to translate).

Comment (ivanseidel) [reply examples/docs/en-US/table.md]: I did manage to add Spanish, but couldn't find the right spot on Chinese, can you give me a hand on that?

**Comment by ivanseidel** [comment]
Hi @ziyoung can you check those changes?

File: examples/docs/en-US/table.md (hunk 1 above)
Comment (ziyoung) [reply examples/docs/en-US/table.md]: https://github.com/ElemeFE/element/blame/dev/examples/docs/zh-CN/table.md#L1540

File: packages/table/src/table-column.js [hunk 2]
```diff
@@ -294,6 +294,11 @@ export default {
       }
     });
 
+    // Deprecation warning for renderHeader property
+    if (this.renderHeader) {
+      console.warn('[Element Warn][Table Column]\'render-header\' property will be deprecated in next major version. please use scoped-slot \'header\' instead.');
```
[the remark this answers is not in the document]
Comment (ziyoung) [reply packages/table/src/table-column.js]: Maybe we can't deprecate it. I find `render-header` is widely used.😂

**Comment by ziyoung** [comment]
You also need to add `Table-column Slot` at the bottom of the document page.

**Comment by ivanseidel** [comment]
Done @ziyoung! All done I guess

File: packages/table/src/table-column.js [hunk 3]
```diff
@@ -444,6 +449,10 @@ export default {
       columnIndex = [].indexOf.call(parent.$el.children, this.$el);
     }
 
+    if (this.$scopedSlots.header) {
```
Comment (ziyoung) [inline packages/table/src/table-column.js]: For selection column, header scope-slot is not allowed to set. When users use this slot, they should get warning information.

File: packages/table/src/table-column.js (hunk 2 above)
Comment (ziyoung) [inline packages/table/src/table-column.js]: I find `render-header` is widely used. Maybe we can't deprecate this attribute in next major version. We should give users enough time to upgrade. 
A potential warning info is like this:
> Comparing to `render-header`, scoped-slot header is easier to use. We recommend users to use scoped-slot header.

File: examples/docs/en-US/table.md [hunk 4]
```diff
@@ -1500,6 +1501,79 @@ Customize table column so it can be integrated with other components.
...
+    style="width: 100%">
+    <el-table-column
+      label="Date"
+      prop="date">
+    </el-table-column>
+    <el-table-column
+      label="Name"
+      prop="name">
+    </el-table-column>
+    <el-table-column
```
Comment (ziyoung) [inline examples/docs/en-US/table.md]: I find this column is redundant.

File: examples/docs/en-US/table.md [hunk 5]
```diff
@@ -1500,6 +1501,79 @@ Customize table column so it can be integrated with other components.
...
+          @click="handleDelete(scope.$index, scope.row)">Delete</el-button>
+      </template>
+    </el-table-column>
+  </el-table>
+</template>
+
+<script>
+  export default {
+    data() {
+      return {
```
Comment (ziyoung) [inline examples/docs/en-US/table.md]: `search` should be included in the return value.

File: examples/docs/en-US/table.md [hunk 6]
```diff
@@ -1500,6 +1501,79 @@ Customize table column so it can be integrated with other components.
...
+          date: '2016-05-04',
+          name: 'Tom',
+          address: 'No. 189, Grove St, Los Angeles'
+        }, {
+          date: '2016-05-01',
+          name: 'Tom',
+          address: 'No. 189, Grove St, Los Angeles'
+        }]
+      }
+    },
```
Comment (ziyoung) [inline examples/docs/en-US/table.md]: You should add missing methods.
```js
methods: {
 handleEdit(){},
 handleDelete(){}
}
```

**Comment by ivanseidel** [comment]
Should be ok now @ziyoung

**Comment by ivanseidel** [comment]
Should be ok now @ziyoung

**Comment by ivanseidel** [comment]
😍😍😍

**Comment by KnorienChang** [comment]
> This is not working in 2.4.9. does anyone could help me ?
> 
> The guide said: Custom content for table header. The scope parameter is { column, $index }
> 
> my code:
> 
> ```
> <el-table-column prop="name" label="Name">
>         <template slot="header" slot-scope="scope">
>           <el-input @input="test(scope)" size="mini" placeholder="Type to search" />
>         </template>
>       </el-table-column>
> ```
> while here scope is type of function, am i do something wrong?

slot-scope="slot"
example:https://github.com/ElemeFE/element/pull/13012/files/991ef98f67d5380c26af6d9bf09c1f7dbda5c7e1#diff-f9a0f43892aff2d5d46d321fa34b903cR1530

**Comment by dhthomas76** [comment]
The documentation for the new header scoped-slot says "Custom content for table header. The scope parameter is { column, $index }", but I am unable to access the column or $index properties.  Whenever I access the "slot" variable it simply returns a render function (not an object with the column and $index properties).  Could someone provide an example of using the new header scoped-slot while accessing the column within the template? 

`<template slot="header" slot-scope="slot">
		       <div>{{slot.column.label}}</div>
</template>`

**Comment by felipe-augusto** [comment]
@Flex-Xuan @KnorienChang @dhthomas76 This issue has been fixed on PR #13263, but has not been released yet.

**Comment by Flex-Xuan** [comment]
> @Flex-Xuan @KnorienChang @dhthomas76 This issue has been fixed on PR #13263, but has not been released yet.

Got it, thanks

**Comment by yolofit** [comment]
#21320

**Pull request merged**